### PR TITLE
[ui-components] Typescript build

### LINKED ISSUE #53: [ui-components] create top-level module declaration
In order to be imported by external Typescript projects, `ui-components` needs a module declaration file to be included in the output bundle.  Relevant links and discussions:
- [publishing declarations in your NPM package](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package)
- generate per-component declaration files by setting `declaration: true` in [TS compiler options](https://www.typescriptlang.org/docs/handbook/compiler-options.html))
- discussion re: [proposal for supporting declarations bundling](https://github.com/microsoft/TypeScript/issues/4433)
- a couple options for external packages that could potentially help ([1](https://github.com/timocov/dts-bundle-generator), [2](https://github.com/Swatinem/rollup-plugin-dts))

### PR DESCRIPTION
**Pull request by nathanstilwell**
# TypeScript support for Design System components

## Building Typescript declarations with `tsc`

Currently the `ui-components` package is authored in Typescript and then transpiled to JavaScript using Babel with the TypeScript preset. This has two effects: first, that the code is never type checked (we leave that to the IDE environment) and second, declarations are not generated for the transpiled code. This results in functioning package code that can't be typed by the consuming application. (see [here](https://iamturns.com/typescript-babel/) for more details)

I am setting up the TypeScript compiler in `ui-components` to generate declarations so consuming applications will have type definitions. This has the added benefit of adding type checking to the build process. To do this the build now takes place in two npm scripts, `build:bundle` (running `webapck`) and `build typescript` (running the `tsc`). This continues to generate a `dist/main.js` (and a source map at `dist/main.js.map`), but will now additionally place TypeScript declarations in `dist/types`. While the type declarations are broken out into a directory structure, they are rolled up into an `index.d.ts` that is noted in the `package.json` for consuming applications. 

## Adding types

Now that we have type support and have set all our `noImplicitAny`s to `true`, you will notice there are also changes to the components themselves to remove errors and warnings from the TypeScript compiler. There were also some minor adjustments made to how components are exported including their types and interfaces for testing and catalog story purposes.

## A note about Webpack

In doing this work I am less and less convinced that Webpack in an appropriate tool for bundling our library code. It adds quite a lot of bulk and obfuscates the code to an unnecessary degree (debugging is alleviated with source maps). This has great value in an application setting where a single bundle is desired, but considering we are bundling a package to be further bundled by an application it feel inefficient. Webpack is a a powerful tool due to it's _loader_ architecture that helps us combine considerations of TypeScript, ESNext features, Sass, and Astroturf, but I can't help but sense there is a better solution. I only note this here to point out that I **do not** seek to solve this problem in this pull request, but will continue to consider other tools that may work better for the task.

fixes #53, #61

### CODE REVIEW

File: packages/storybook-ui-components/tsconfig.json [hunk 1]
```diff
@@ -13,47 +13,7 @@
...
-      As a starting point for work that builds on the integration of Typescript into our Storybook
-      setup, there are a couple packages that attempt to tackle this and seem worth investigating:
-      - https://github.com/timocov/dts-bundle-generator
-      - https://github.com/Swatinem/rollup-plugin-dts
-
-      In the interest of not getting completely bogged-down I've opted to punt on the aforementioned
-      and have filed an issue (https://github.com/cockroachdb/ui/issues/53) to address this.
-    */
-    "noImplicitAny": false,
-
```
Comment (j-low) [inline packages/storybook-ui-components/tsconfig.json:56]: Beautiful seeing this go away

Comment (nathanstilwell) [reply packages/storybook-ui-components/tsconfig.json:56]: 🙌

File: yarn.lock [hunk 2]
```diff
@@ -0,0 +1,5096 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (laurenbarker) [inline yarn.lock:1]: Isn't this the lockfile that the `--no-lockfile` flag is saying don't use?

File: packages/icons/package.json [hunk 3]
```diff
@@ -13,9 +13,9 @@
     "build:generate": "svgr --typescript --icon -d src/components svg",
     "build:watch": "tsc --watch",
     "clean": "rimraf dist src/components",
-    "prepublishOnly": "npm-run-all clean build",
+    "prepublish": "npm-run-all clean build",
     "test": "node notests.js",
-    "start": "npm run build:watch"
+    "start": "npm-run-all build:watch"
```
[the remark this answers is not in the document]
Comment (nathanstilwell) [reply packages/icons/package.json:18]: _run-all_ just makes it easier to add tasks to the end of the task. I can take it out if it bothers you.

File: package.json [hunk 4]
```diff
@@ -5,7 +5,7 @@
     "lerna": "^3.20.2"
   },
   "scripts": {
-    "init": "yarn && lerna bootstrap --force-local && lerna link",
+    "init": "yarn install --no-lockfile && lerna bootstrap --force-local && lerna link",
```
[the remark this answers is not in the document]
Comment (nathanstilwell) [reply package.json:8]: I believe `--no-lockfile` means to not update the lockfile when running install. My thinking here is that when someone is checking out changes or just getting the repo running they may not want to change the lockfile and dirty the repo. So the install will still use the lockfile to ensure that everyone is on the same vesion, but updating the lockfile should be intentional.

[the remark this answers is not in the document]
Comment (nathanstilwell) [reply package.json:8]: @laurenbarker  On second glance, I think you are right. `--no-lockfile` means don't use a lockfile. I think I wanted `--pure-lockfile` which doesn't update the lockfile. I'll put up another PR.

**Pull request merged**